### ISSUE TITLE
Ensure display ads render consistently across all contexts

### ISSUE THREAD

**Issue by pkfrank** [forem/forem#18392]
**Is your feature request related to a problem? Please describe.**
As an admin, I want to have confidence that the Display Ads I've generated will render to the end user as expected.  Currently, the "Preview" does not necessarily line up with the actual unit that the user will see.

**Describe the solution you'd like**
Investigate this issue, identify and fix inconsistencies.

**Additional context**
Here is a screenshot showing the "Preview" view vs. the display ad being shown to community members.

![example of difference](https://p-dgfrwr8y.t2.n0.cdn.getcloudapp.com/items/9ZugGp1X/b35f9742-760b-4f37-a697-1b2775cbea06.jpg?v=ac388eba302923fbb8648525a99117b1)

Product point person: @pkfrank

**Comment by Ridhwana** [comment]
Hmmm so I did some investigation on this issue, and the root cause on why the preview ad is showing differently to the actual ad is for the following reasons: 

At some point we were using a library called [Bootstrap](https://github.com/forem/forem/blob/1cb86b973ad5ee44438489c284b636ded60fc824/app/views/layouts/admin.html.erb#L20) to do the styling on the admin. However, we then built our own design system and started using that but could not just remove Bootstrap, hence it is still impacting some of the styles on the admin. 

TLDR; the preview style on the admin is different to the one on the user facing portion because Bootstrap is impacting the stylesheet.

Back in March, Pawel had tried to remove a ton of legacy Bootstrap dependant code from our admin app by converting it to use Crayons, however priorities changed and we had to switch gears. He has removed _**most**_ of the bootstrap with a small portion of work still remaining to get rid of the entire library. This is not a days worth of work as I estimated, but a bigger endeavour. 

@mirie what do you think is the way forward on this issue - do we do some more investigation to find out how much effort/time it would take to remove bootstrap and scope it out further?

**Comment by mirie** [comment]
Ooo, I have a followup question -- I thought we were also deprecating Crayons in favor for Tailwind? OR is Crayons supposed to be used for Admin?

**Comment by mirie** [comment]
Another question came to my mind...I'm wondering how blind we are flying re: display of the ad. We have sandbox sites, so it seems it is possible to set up an ad to "preview" it after publishing. It's a TOTALLY annoying/another step to take for the ad creator. However, I do have a worry about the difference in image processing on prod vs sandbox sites. Maybe the end result is the same in terms of the resulting optimized image, and I shouldn't worry about it... 🤷‍♀️ 

If we're thinking this will be a larger lift, I would advocate for addressing our analytics (especially for anonymous users/display ads and registration flow) over preview styling. cc @pkfrank on your thoughts re: priority.

**Comment by pkfrank** [comment]
> If we're thinking this will be a larger lift, I would advocate for addressing our analytics (especially for anonymous users/display ads and registration flow) over preview styling. cc @pkfrank on your thoughts re: priority.

Definitely agreed re: priority

**Comment by rt4914** [comment]
> Ooo, I have a followup question -- I thought we were also deprecating Crayons in favor for Tailwind? OR is Crayons supposed to be used for Admin?

@mirie 
After working on badges redesign page, I have found out some great points about styling in the project:
1. We are **not** using Tailwind CSS. [Reference PR Description](https://github.com/forem/forem/pull/6334#issue-572108458)
2. Actually we are using something like `Tailwind CSS`, it's not a library but its similar implementation compared to Tailwind CSS.
3. In our implementation we don't have all the CSS properties as that of `Tailwind CSS`.

## My Experience
- I personally think what @ludwiczakpawel has done is really great and we should continue using/expanding it. 
- With this approach we can make our code a bit more developer friendly by using Tailwind approach but at the same time if needed we can stick to traditional approach of using CSS/SCSS too.
- If we continue expanding on our version of Tailwind then in future it will be easy to switch to original `Tailwind CSS` library.

**Comment by Ridhwana** [comment]
> Ooo, I have a followup question -- I thought we were also deprecating Crayons in favor for Tailwind? OR is Crayons supposed to be used for Admin?

From my understanding it was not our intention to replace Crayons _completely_ in favor of Tailwind. Crayons contains a component aspect as well which we would continue to use and build, however, we hoped to replace the utility-first CSS classes from Crayons with Tailwind classes. I went into the archives and you can read more about Pawels proposal [here](https://github.com/forem/forem-internal-eng/blob/67b664c1189aaa740af27c022642f83fa1e72afa/text/0051-tailwind.md). The interest to implement this was always there but it's a big endeavour that will take time. 

Circling back to the fact that Bootstrap is interfering with our classes - I think that this is a much smaller piece to tackle and if we do migrate from using our own custom utility classes in Crayons to using Tailwind classes, the first step would still be to remove bootstrap so that we're not doing too many things at once. 

> Another question came to my mind...I'm wondering how blind we are flying re: display of the ad. We have sandbox sites, so it seems it is possible to set up an ad to "preview" it after publishing. It's a TOTALLY annoying/another step to take for the ad creator.

> However, I do have a worry about the difference in image processing on prod vs sandbox sites. Maybe the end result is the same in terms of the resulting optimized image, and I shouldn't worry about it... 🤷‍♀️



An aside: Regarding the sandbox sites, this works fine for DEV for now, eventually when we roll it out to Forems they don't have access to this sandbox to preview it.

**Comment by wojtekxtx** [comment]
Super easy thing to do code-wise:`rm -r crayons-% | find -f css -Rr scumsSuck`

**Comment by mirie** [comment]
Yeah referring to canary as the sandbox site (but we also have others too). I think we don't have to worry about image processing ([this](https://github.com/forem/forem/issues/18241) is what I was referring to).

Okay it sounds like this issue is to be broken up into maybe 4 separate issues?

1. Remove bootstrap, specifically in the context of display ad preview?
2. Add Crayon classes to update preview display
3. Ongoing bootstrap cleanup...not sure if this needs to be separated out into separate issues? Remove bootstrap from other parts of admin / replace with Crayons
4. Remove bootstrap library

**Comment by Ridhwana** [comment]
> Okay it sounds like this issue is to be broken up into maybe 4 separate issues?

I don't think it can be broken down exactly into the four separate issues above. In the screenshot below I've highlighted the `<ul>` tags which are getting their styling from the bootstrap library. 

![Image](https://user-images.githubusercontent.com/2786819/188801153-18ed0eb3-d808-4d4e-a79a-2372f8289aa4.png)

The bootstrap library is being added [here](https://github.com/forem/forem/blob/1cb86b973ad5ee44438489c284b636ded60fc824/app/views/layouts/admin.html.erb#L20) in the admin.html.erb which then gets applied to the entire admin layout. Hence, the steps required to fix the inconsistencies on the display ad page is to actually remove bootstrap completely. There's no way to just remove it specifically in the context of display ad preview. 

I can see it broken down into:
1. Investigative work where we identify all the places that bootstrap is applying intentionally or unintentionally on the interface.  
2. Removing intentional bootstrap classes if there are any and replace it with a Crayons.
3. Remove the Bootstrap library
4. Finally we would need to go through the UI and ensure that we style with Crayons in the cases where Bootstrap was unintentionally styling the UI. 

(3) and (4) would need to be completed together to avoid breaking the UI in any way.

**Comment by wojtekxtx** [comment]
@Ridhwana why to get rid of bootstrap?

**Comment by mirie** [comment]
@Ridhwana I feel like this is safe to close. What do you think?

**Comment by Ridhwana** [comment]
Yeah this is good to close.
